kernel/src/hil/touch.rs

```diff
@@ -0,0 +1,97 @@
+//! Interface for touch input devices
+
+use crate::ReturnCode;
+
+/// Touch Event Status
+#[derive(Debug, Copy, Clone)]
+pub enum TouchStatus {
+    Pressed,
+    Released,
+}
+
+/// Gesture event
+#[derive(Debug, Copy, Clone)]
+pub enum GestureEvent {
+    SwipeUp,
+    SwipeDown,
+    SwipeLeft,
+    SwipeRight,
+    ZoomIn,
+    ZoomOut,
+}
+
+/// A single touch event's data
+#[derive(Copy, Clone)]
+pub struct TouchEvent {
+    pub status: TouchStatus,
+    /// touch (x, y) position
+    pub x: usize,
+    pub y: usize,
+
+    /// touch id, value defined by the driver
+    pub id: usize,
+
+    /// touch area
+    pub area: Option<usize>,
+
+    /// touch weight
+    pub weight: Option<usize>,
+}
+
+/// Single touch panels should implement this
+pub trait Touch {
+    /// Enable the touche panel
+    fn enable(&self) -> ReturnCode;
+
+    /// Disable the touch panel
+    fn disable(&self) -> ReturnCode;
+
+    /// Set the touch client
+    fn set_client(&self, touch_client: &'static dyn TouchClient);
+}
+
+/// Multi-touch panels should implement this
+pub trait MultiTouch {
+    /// Enable the touche panel
+    fn enable(&self) -> ReturnCode;
+
+    /// Disable the touch panel
+    fn disable(&self) -> ReturnCode;
+
+    /// Returns the number of concurently supported touches
+    /// This function must be called in the same interrupt
+    /// as the event, otherwise data might not be available.
+    fn get_num_touches(&self) -> usize;
```

Comment:
I'm not sure I follow this API. Will the number of concurrent touches supported change? What is the event? Doesn't the event report how many touches there were?

Reply:
Each touch screen might support a different number of touches, I expect this function to be constant. Maybe add const fn?

Reply:
This is not necessarily a suggestion, but the way we handle something similar to this in the time HIL is by using an associated type to denote the clock frequency. The associated type itself has only a static method (i.e., one that doesn't take a `self`) and therefore, presumably, is easier for the compiler to resolve statically and inline the constant.

But... I don't know that we ever evaluated that it actually makes a difference, and I don't know if others agree that it's a better interface than just having an instance method in the trait.

Reply:
I get your point, but it does add some complexity to the code and seems to make it more difficult to read.